File: app/src/main/java/me/ykrank/s1next/view/internal/ToolbarDelegate.java
package me.ykrank.s1next.view.internal;

import android.support.annotation.Nullable;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.PopupWindow;
import android.widget.TextView;

import com.google.common.base.Preconditions;

import me.ykrank.s1next.R;
import me.ykrank.s1next.databinding.PopWindowTitleBinding;
import me.ykrank.s1next.util.ResourceUtil;

/**
 * This class represents a delegate which you can use to add
 * {@link Toolbar} to {@link AppCompatActivity}.
 */
public final class ToolbarDelegate {

    private final AppCompatActivity mAppCompatActivity;
    private final Toolbar mToolbar;
    @Nullable
    private TextView longTitleView;

    public ToolbarDelegate(AppCompatActivity appCompatActivity, Toolbar toolbar) {
        this.mAppCompatActivity = appCompatActivity;
        this.mToolbar = toolbar;

        setUpToolbar();
    }

    /**
     * Sets a {@link android.widget.Toolbar Toolbar} to act as the {@link android.support.v7.app.ActionBar}
     * for this Activity window.
     * Also displays home as an "up" affordance in Toolbar.
     */
    private void setUpToolbar() {
        // designate a Toolbar as the ActionBar
        mAppCompatActivity.setSupportActionBar(mToolbar);
        Preconditions.checkNotNull(mAppCompatActivity.getSupportActionBar())
                .setDisplayHomeAsUpEnabled(true);
        //self defined title TextView
        longTitleView = (TextView) mToolbar.findViewById(R.id.toolbar_title_long);
        if (longTitleView != null) {
            longTitleView.setOnLongClickListener(v -> onLongClick(v, longTitleView.getText().toString()));
        }
    }

    /**
     * Sets Toolbar's navigation icon to cross.
     */
    public void setupNavCrossIcon() {
        mToolbar.setNavigationIcon(ResourceUtil.getResourceId(mAppCompatActivity.getTheme(),
                R.attr.iconClose));
    }

    public Toolbar getToolbar() {
        return mToolbar;
    }

    /**
     * Set title
     *
     * @return whether handle this action
     */
    public boolean setTitle(CharSequence title) {
        if (longTitleView != null && !TextUtils.equals(longTitleView.getText(), title)) {
            mToolbar.setTitle(null);
            longTitleView.setText(title);
            return true;
        }
        return false;
    }

    private boolean onLongClick(View anchor, String title) {
        PopWindowTitleBinding binding = PopWindowTitleBinding.inflate(LayoutInflater.from(anchor.getContext()));
        binding.setTitle(title);
        final PopupWindow popupWindow = new PopupWindow(binding.getRoot(),
                ViewGroup.LayoutParams.WRAP_CONTENT, ViewGroup.LayoutParams.WRAP_CONTENT, true);
        popupWindow.setOutsideTouchable(true);
        popupWindow.setFocusable(false);
        popupWindow.showAsDropDown(anchor);
        return true;
    }
}

File: app/src/main/java/me/ykrank/s1next/view/internal/ToolbarDelegate.kt
package me.ykrank.s1next.view.internal

import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.Toolbar
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.PopupWindow
import android.widget.TextView
import com.github.ykrank.androidtools.extension.toast
import com.github.ykrank.androidtools.util.ClipboardUtil
import com.github.ykrank.androidtools.util.ResourceUtil
import me.ykrank.s1next.R
import me.ykrank.s1next.databinding.PopWindowTitleBinding

/**
 * This class represents a delegate which you can use to add
 * [Toolbar] to [AppCompatActivity].
 */
class ToolbarDelegate(private val mAppCompatActivity: AppCompatActivity, val toolbar: Toolbar) {
    private var longTitleView: TextView? = null

    init {
        setUpToolbar()
    }

    /**
     * Sets a [Toolbar][android.widget.Toolbar] to act as the [android.support.v7.app.ActionBar]
     * for this Activity window.
     * Also displays home as an "up" affordance in Toolbar.
     */
    private fun setUpToolbar() {
        // designate a Toolbar as the ActionBar
        mAppCompatActivity.setSupportActionBar(toolbar)
        mAppCompatActivity.supportActionBar?.setDisplayHomeAsUpEnabled(true)
        //self defined title TextView
        longTitleView = toolbar.findViewById(R.id.toolbar_title_long)
        longTitleView?.setOnLongClickListener { v -> onLongClick(v, longTitleView?.text.toString()) }
    }

    /**
     * Sets Toolbar's navigation icon to cross.
     */
    fun setupNavCrossIcon() {
        toolbar.setNavigationIcon(ResourceUtil.getResourceId(mAppCompatActivity.theme,
                R.attr.iconClose))
    }

    /**
     * Set title
     *
     * @return whether handle this action
     */
    fun setTitle(title: CharSequence?): Boolean {
        longTitleView?.apply {
            if (!toolbar.title.isNullOrEmpty()) {
                toolbar.title = null
            }
            if (!title.isNullOrEmpty() && this.text != title) {
                this.text = title
            }
            return true
        }
        return false
    }

    private fun onLongClick(anchor: View, title: String): Boolean {
        val binding = PopWindowTitleBinding.inflate(LayoutInflater.from(anchor.context))
        binding.title = title
        binding.root.setOnClickListener { v ->
            ClipboardUtil.copyText(v.context, title, title)
            v.context.toast(R.string.title_copied)
        }

        val popupWindow = PopupWindow(binding.root,
                ViewGroup.LayoutParams.WRAP_CONTENT, ViewGroup.LayoutParams.WRAP_CONTENT, true)
        popupWindow.isOutsideTouchable = true
        popupWindow.isFocusable = false
        popupWindow.showAsDropDown(anchor)
        return true
    }
}
